### PR TITLE
 Updated: simulator api endpoints implemented in /api/v1/projects enpoint 

### LINKED ISSUE #3778: Creation of new api endpoints in the api/v1 format part-1
### Tasks to be done

- [x] Create new api endpoint to fetch user information , check project access , get project details
- [x] Create new api to get project data
- [x] Create new api to update project data

### Are you working on this?

Yes

### PR DESCRIPTION
**Pull request by Arnabdaz**
Fixes #3778 

#### Describe the changes you have made in this PR -
Added new api endpoints:

1. check_edit_access:
GET `/api/v1/projects/:id/check_edit_access`
- checks edit access of a user for the specified project id
- if access is granted user Info is returned
- if user does not have access 403 returned
- if user is not signed in 401 is returned
 
2. circuit_data:
`GET /api/v1/projects/:id/circuit_data`
- returns the project's circuit data if available & user has view access to project
- return 403 if user does not have view access of project

3. update_circuit:
`PATCH /api/v1/projects/update_circuit`
- update circuit data if user has edit access
- return 403 if user does not have edit access
- return 401 if user is not signed in 

4. create:
`POST /api/v1/projects`
- create new project if user is singed in
- return 401 if user is not signed in

### Screenshots of the changes (If any) -

image
image
image
image

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by Arnabdaz** [comment]
cc @tachyons

**Comment by tachyons** [comment]
CI is failing due to insufficient coverage

**Comment by Arnabdaz** [comment]
cc @tachyons

File: app/controllers/api/v1/projects_controller.rb [hunk 1]
```diff
@@ -111,6 +180,30 @@ def set_project
       end
     end
 
+    # Update circuit data Related methods
+
+    # FIXME: remove this logic after fixing production data
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:185]: What is wrong with production data ?

Comment (Arnabdaz) [reply app/controllers/api/v1/projects_controller.rb:185]: I actually it was in the old simulator controller also 
if only `current_user.projects.friendly.find_by(id: params[:id])` is used it does not allow collebrator access 
thats why `current_user.projects.friendly.find_by(id: params[:id]) || Project.friendly.find(params[:id])` is being used currently 

we can update it to 
```
def set_user_project
  @project = current_user.projects.friendly.find_by(id: params[:id])

  if @project.nil?
    @project = Project.friendly.find(params[:id])
    unless @project.collaborators.include?(current_user)
      render json: { error: 'You do not have access to this project' }, status: :unauthorized
      return
    end
  end
end
```

File: app/controllers/api/v1/projects_controller.rb [hunk 2]
```diff
@@ -111,6 +180,30 @@ def set_project
...
+      @image_file = return_image_file(params[:image])
+      @project.image_preview = @image_file
+      @project.name = sanitize(params[:name])
+    end
+
+    def handle_image_file_cleanup
+      @image_file.close
+      File.delete(@image_file) if check_to_delete(params[:image])
+    end
+    ######################
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:205]:
```suggestion

```

File: app/controllers/api/v1/simulator_controller.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+class Api::V1::SimulatorController < Api::V1::BaseController
+  # POST /api/v1/simulator/post_issue
+  def post_issue
+    issue_circuit_data = IssueCircuitDatum.new(data: params[:circuit_data]).tap(&:save)
```
Comment (tachyons) [inline app/controllers/api/v1/simulator_controller.rb:6]:
```suggestion
    issue_circuit_data = IssueCircuitDatum.create(data: params[:circuit_data])
```

File: app/controllers/api/v1/simulator_controller.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    circuit_data_url = "#{request.base_url}/simulator/issue_circuit_data/#{issue_circuit_data.id}"
+    text = "#{params[:text]}\nCircuit Data: #{circuit_data_url}"
+
+    url = ENV.fetch("SLACK_ISSUE_HOOK_URL", nil)
+
+    if url.nil? || !url.start_with?("http://", "https://")
+      render json: { error: "Invalid or missing Slack webhook URL" }, status: :internal_server_error and return
```
Comment (tachyons) [inline app/controllers/api/v1/simulator_controller.rb:14]:
```suggestion
      render json: { error: "Invalid or missing Slack webhook URL" }, status: 422 and return
```

File: app/controllers/api/v1/simulator_controller.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    response = HTTP.post(url, json: { text: text })
+    unless response.code == 200
+      render json: { error: "Failed to submit issue to Slack" }, status: :internal_server_error and return
```
Comment (tachyons) [inline app/controllers/api/v1/simulator_controller.rb:19]:
```suggestion
      render json: { error: "Failed to submit issue to Slack" }, status: 422 and return
```

File: spec/factories/project_data.rb [hunk 6]
```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 
 FactoryBot.define do
-  factory :project_datum do
+  factory :project_datum, class: "ProjectDatum" do
+    data { { name: "circuit data" }.to_json }
```
Comment (tachyons) [inline spec/factories/project_data.rb:5]:
```suggestion
    data { { name: "circuit data" }}
```

**Review by tachyons — CHANGES_REQUESTED** [review]

File: spec/requests/api/v1/projects_controller/circuit_data_spec.rb [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    let!(:empty_project) { FactoryBot.create(:project, project_access_type: "Public", author: user) }
+
+    context "when unauthenticated user fetches public project circuit data" do
+      before do
+        get "/api/v1/projects/#{public_project.id}/circuit_data", as: :json
+      end
+
+      it "returns circuit data" do
+        expect(response).to have_http_status(:ok)
+        parsed_response = eval(response.body).with_indifferent_access
```
[the remark this answers is not in the document]
Comment (tachyons) [reply spec/requests/api/v1/projects_controller/circuit_data_spec.rb:28]: use response.parsed_body

**Comment by Arnabdaz** [comment]
@tachyons can please you review

**Pull request merged**